File: src/lib/jimeng.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { GeneratedImage, Usage, ImageError } from "@/lib/types";

export interface GenerateImageParams {
  model: string;
  prompt: string;
  image?: string[];
  size?: string;
  /** 仅当需要组图时设置；不要传 "disabled" */
  sequential?: "auto";
  maxImages?: number;
  layerDecomposition?: boolean;
  /**
   * 是否在生成图右下角加 "AI生成" 水印。
   * 留空走 Y-agent 默认（不加）；Seedream API 默认 true，Y-agent 后端会显式补 false 关闭。
   * 如未来需要 UI 开关，往这里传 true/false 即可。
   */
  watermark?: boolean;
  /** 输出文件格式：png | jpeg。仅 5.0 Pro / 5.0 Lite 支持。 */
  outputFormat?: "png" | "jpeg";
  /** 工具调用。当前仅支持 web_search（5.0 Lite 专属）。 */
  tools?: Array<"web_search">;
  /** 提示词优化模式：standard | fast。fast 仅 5.0 Pro / 4.0 支持。 */
  optimizePromptMode?: "standard" | "fast";
  /** 背景透明：transparent | opaque。仅 5.0 Pro 专属。 */
  background?: "transparent" | "opaque";
}

export interface GenerateImageResponse {
  images: GeneratedImage[];
  /** 是否走 demo 模式（不消耗 token） */
  isDemo: boolean;
  /** API 顶层 usage（生成张数 / token / 工具调用次数） */
  usage?: Usage;
  /** API 顶层 error（整个请求都失败时） */
  error?: ImageError;
  /** P4：实际输出格式（png|jpeg），仅 5.0 Pro / Lite 返回 */
  output_format?: "png" | "jpeg";
}

export async function generateImage(params: GenerateImageParams): Promise<GenerateImageResponse> {
  return await invoke<GenerateImageResponse>("jimeng_generate", {
    params: applyOutputFormatDefault(params),
  });
}

// ============================================================================
// P1：流式生图
// ============================================================================

/** 流事件类型（与 Rust 端 StreamEvent 对齐）
 * P5: partial_image / partial_image_b64 可能带 `localPath`（24h 兜底用）。
 * Rust 端先 emit 一次 localPath 为空的事件让前端立刻入库；下载完成后再 emit 一次带 localPath。
 * 前端应以后续事件里的 localPath 为准，更新已入库资产的 payload.localPaths。
 *
 * P6+：`output_format` 由 Rust 端从入参 `params.output_format` 推算后透传过来。
 * SSE partial 事件本身不带此字段（它是顶层字段，在 completed 事件里），但 partial
 * 资产入库时需要它。5.0 系列 = Some("png"|"jpeg")，4.5/4.0 = None。
 */
export type StreamEvent =
  | {
      type: "partial_image";
      request_id: string;
      index: number;
      url: string;
      size?: string;
      localPath?: string;
      output_format?: string;
    }
  | {
      type: "partial_image_b64";
      request_id: string;
      index: number;
      b64: string;
      localPath?: string;
      output_format?: string;
    }
  | { type: "partial_failed"; request_id: string; index?: number; code?: string; message?: string }
  | { type: "completed"; request_id: string; usage?: Usage }
  | { type: "aborted"; request_id: string; reason: string };

export interface StreamHandlers {
  /** 单图完成（URL 直链） */
  onPartial?: (img: {
    index: number;
    url: string;
    size?: string;
    localPath?: string;
    /** Rust 端从入参推算的 output_format（5.0 系列有值，4.5/4.0 = undefined） */
    outputFormat?: string;
  }) => void;
  /** 单图完成（base64，自动补 data:image/png;base64, 前缀） */
  onPartialB64?: (img: {
    index: number;
    url: string;
    localPath?: string;
    outputFormat?: string;
  }) => void;
  /** 单图失败 */
  onPartialFailed?: (info: { index?: number; code?: string; message?: string }) => void;
  /** 全部完成 */
  onCompleted?: (info: { usage?: Usage }) => void;
  /** 流中断（InternalServiceError / 网络断开 / 顶层 error） */
  onAborted?: (info: { reason: string }) => void;
}

/**
 * 启动一次流式生图，立即返回 request_id。
 * 后台通过 Tauri event `jimeng://stream` 持续推事件，handlers 会被对应回调命中。
 * InternalServiceError 或顶层 error 会在 `onAborted` 里告知，**不**抛异常。
 */
export async function generateImageStream(
  params: GenerateImageParams,
  handlers: StreamHandlers = {}
): Promise<string> {
  const requestId = await invoke<string>("jimeng_generate_stream", {
    params: applyOutputFormatDefault(params),
  });

  // 监听本次 request_id 的事件（用事件 payload 里的 request_id 过滤，避免与其它并发请求串台）
  let finished = false;
  const safeUnlisten = () => {
    try {
      unlisten();
    } catch {
      /* ignore */
    }
  };
  const unlisten: UnlistenFn = await listen<StreamEvent>("jimeng://stream", (e) => {
    const ev = e.payload;
    if (ev.request_id !== requestId) return;
    switch (ev.type) {
      case "partial_image":
        handlers.onPartial?.({
          index: ev.index,
          url: ev.url,
          size: ev.size,
          localPath: ev.localPath,
          outputFormat: ev.output_format,
        });
        break;
      case "partial_image_b64": {
        // 兼容 data:image/... 前缀与裸 base64
        const b64 = ev.b64;
        const url = b64.startsWith("data:") ? b64 : `data:image/png;base64,${b64}`;
        handlers.onPartialB64?.({
          index: ev.index,
          url,
          localPath: ev.localPath,
          outputFormat: ev.output_format,
        });
        break;
      }
      case "partial_failed":
        handlers.onPartialFailed?.({
          index: ev.index,
          code: ev.code,
          message: ev.message,
        });
        break;
      case "completed":
        if (finished) return;
        finished = true;
        handlers.onCompleted?.({ usage: ev.usage });
        // 收尾后释放本次 listener
        safeUnlisten();
        break;
      case "aborted":
        if (finished) return;
        finished = true;
        handlers.onAborted?.({ reason: ev.reason });
        safeUnlisten();
        break;
    }
  });

  // 防御性兜底：10 分钟内都没收到 completed/aborted（理论 Rust 端会 bail），主动释放。
  // 避免 listener 永远挂着。
  setTimeout(
    () => {
      if (finished) return;
      finished = true;
      safeUnlisten();
    },
    10 * 60 * 1000
  );

  return requestId;
}

/** 把后端原始错误信息转成更友好的提示 */
export async function explainError(raw: string): Promise<string> {
  return await invoke<string>("explain_error", { raw });
}

/**
 * 给 outputFormat 设默认值：Y-agent 业务规则 — 5.0 Pro / Lite 默认 PNG，4.5 / 4.0 不设（API 默认 jpeg）。
 * - 调用方显式传了 png/jpeg → 不动
 * - 调用方传 undefined：
 *   - 5.0 Pro / Lite → 设 "png"
 *   - 4.5 / 4.0 → 不设（保持 undefined，让 API 走默认 jpeg）
 *   - 其它/未知 → 不设
 *
 * Export 给入库路径用：partial 阶段还没拿到 API 响应，必须用归一化后的
 * 请求参数写入 payload.outputFormat，否则 5.0 Lite 默认场景下 partial 资产
 * 永远没格式字段，详情页"格式"会显示 "—"。
 * （jimeng_generate / jimeng_generate_stream 内部也调，重复调用是幂等的。）
 */
export function applyOutputFormatDefault(
  params: GenerateImageParams
): GenerateImageParams {
  if (params.outputFormat) return params;
  const id = params.model;
  const is5x =
    id.startsWith("doubao-seedream-5-0-pro") ||
    id.startsWith("doubao-seedream-5-0-lite") ||
    id.startsWith("doubao-seedream-5-0-260128");
  if (is5x) {
    return { ...params, outputFormat: "png" };
  }
  return params;
}

File: src-tauri/src/jimeng.rs
use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use serde::{Deserialize, Serialize};

const ENDPOINT: &str = "https://ark.cn-beijing.volces.com/api/v3/images/generations";

/// 5.0 Pro 不支持的参数组合：调用前在 Rust 端做护栏，避免回 400 时还要走 explain_error。
fn validate_params(params: &GenerateImageParams) -> Result<(), String> {
    let is_5_0_pro = params.model == "doubao-seedream-5-0-pro-260628";
    let is_4_5_or_4_0 = matches!(
        params.model.as_str(),
        "doubao-seedream-4-5-251128" | "doubao-seedream-4-0-250828"
    );
    if is_5_0_pro && params.sequential_image_generation.is_some() {
        return Err(
            "InvalidParameter: 5.0 Pro 不支持组图（sequential_image_generation）".into(),
        );
    }
    if let Some(fmt) = &params.output_format {
        if is_4_5_or_4_0 && fmt == "png" {
            return Err(
                "InvalidParameter: 该模型仅支持 jpeg 输出（output_format=png 不可用）".into(),
            );
        }
    }
    if let Some(opts) = &params.optimize_prompt_options {
        if let Some(mode) = &opts.mode {
            if mode == "fast"
                && !matches!(
                    params.model.as_str(),
                    "doubao-seedream-5-0-pro-260628" | "doubao-seedream-4-0-250828"
                )
            {
                return Err("InvalidParameter: optimize_prompt_options.mode=fast 仅 5.0 Pro / 4.0 支持".into());
            }
        }
    }
    Ok(())
}

/// Demo 模式：API Key 以 "demo-" 开头时触发。
/// 用于在没有真实 Key 的情况下让 UI 走通，节省 token 调试。
fn is_demo_key(api_key: &str) -> bool {
    api_key.starts_with("demo-")
}

fn make_demo_svg(label: &str) -> String {
    let svg = format!(
        r##"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024">
  <defs>
    <linearGradient id="g" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#1a1a20"/>
      <stop offset="100%" stop-color="#2a1a4a"/>
    </linearGradient>
  </defs>
  <rect width="1024" height="1024" fill="url(#g)"/>
  <text x="512" y="480" text-anchor="middle" fill="#7c5cff" font-family="sans-serif" font-size="48" font-weight="600">Y-agent DEMO</text>
  <text x="512" y="540" text-anchor="middle" fill="#a0a0a8" font-family="sans-serif" font-size="22">{label}</text>
  <text x="512" y="600" text-anchor="middle" fill="#6a6a72" font-family="sans-serif" font-size="18">在设置中填入真实即梦 API Key 即可生成真实图像</text>
</svg>"##
    );
    format!("data:image/svg+xml;base64,{}", B64.encode(svg.as_bytes()))
}

#[derive(Debug, Serialize)]
pub struct GenerateImageParams {
    pub model: String,
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sequential_image_generation: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sequential_image_generation_options: Option<SequentialOptions>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub layer_decomposition: Option<bool>,
    /// 是否在生成图右下角加 "AI生成" 水印。
    /// 即梦/豆包 Seedream API 默认 `true`（加水印），与 Y-agent 期望不符。
    /// 调用方可不传；`generate()` 会补 `false`。
    /// 未来要做 UI 开关时，前端传 `true` 进来即可生效。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub watermark: Option<bool>,
    /// 输出文件格式：仅 5.0 Pro / 5.0 Lite 支持 png|jpeg；4.5/4.0 固定 jpeg。
    /// Y-agent 留 None 时后端不传该字段（走 API 默认）。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_format: Option<String>,
    /// 工具调用：当前只支持 `web_search`（5.0 Lite 专属）。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<ToolSpec>>,
    /// 提示词优化模式：standard（默认）|fast（5.0 Pro / 4.0 专属）。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub optimize_prompt_options: Option<OptimizePromptOptions>,
    /// 背景透明：仅 5.0 Pro 专属、图生图场景、PNG 输出。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub background: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SequentialOptions {
    pub max_images: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ToolSpec {
    #[serde(rename = "type")]
    pub kind: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OptimizePromptOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ApiResponse {
    data: Vec<ApiImage>,
    #[serde(default)]
    usage: Option<Usage>,
    #[serde(default)]
    error: Option<ApiError>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ApiError {
    pub code: Option<String>,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Usage {
    #[serde(default, rename = "generated_images")]
    pub generated_images: Option<i32>,
    #[serde(default, rename = "input_images")]
    pub input_images: Option<i32>,
    #[serde(default, rename = "output_tokens")]
    pub output_tokens: Option<i32>,
    #[serde(default, rename = "total_tokens")]
    pub total_tokens: Option<i32>,
    #[serde(default, rename = "tool_usage")]
    pub tool_usage: Option<ToolUsage>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ToolUsage {
    #[serde(default, rename = "web_search")]
    pub web_search: Option<i32>,
}

#[derive(Debug, Deserialize)]
struct ApiImage {
    url: Option<String>,
    #[serde(rename = "b64_json", alias = "b64Json")]
    b64_json: Option<String>,
    /// 实际像素尺寸，如 `2048x2048`（仅 5.0 pro / lite 返回）
    #[serde(default)]
    size: Option<String>,
    /// 输出文件格式（仅 5.0 pro / lite 返回）
    #[serde(rename = "output_format", default)]
    output_format: Option<String>,
    /// 图层拆分场景下返回，底图固定为 0
    #[serde(default, alias = "zIndex")]
    z_index: Option<i32>,
    /// 图层名称
    #[serde(default)]
    name: Option<String>,
    /// 图层描述
    #[serde(default)]
    description: Option<String>,
    /// 图层边界框（图层拆分场景，仅 5.0 pro）
    #[serde(default, rename = "bounding_box")]
    bounding_box: Option<BoundingBox>,
    /// 单图错误（组图场景下某张失败时填）
    #[serde(default)]
    error: Option<ApiError>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct BoundingBox {
    /// 输出底图坐标系绝对像素坐标 [left, top, right, bottom]
    #[serde(default)]
    pub absolute: Option<Vec<i32>>,
    /// 归一化坐标 [left, top, right, bottom]（0-1000）
    #[serde(default)]
    pub normalized: Option<Vec<i32>>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GeneratedImage {
    pub url: String,
    #[serde(default)]
    pub is_base64: bool,
    /// 图层叠放顺序：底图=0，图层从 1 起
    #[serde(skip_serializing_if = "Option::is_none")]
    pub z_index: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// 实际像素尺寸，如 `2048x2048`
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size: Option<String>,
    /// 输出文件格式：png|jpeg
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_format: Option<String>,
    /// 图层边界框（图层拆分场景）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bounding_box: Option<BoundingBox>,
    /// 单图错误（组图场景下某张失败时填）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ApiError>,
}

#[derive(Debug, Serialize)]
pub struct GenerateResult {
    pub images: Vec<GeneratedImage>,
    /// 是否走 demo 模式（前端用这个决定是否入库 demo 资产）
    pub is_demo: bool,
    /// API 顶层 usage（生成张数 / token / 工具调用次数等）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<Usage>,
    /// API 顶层 error（整个请求都没生成图时）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ApiError>,
}

/// 调用即梦图像生成 API。
/// 返回的图片以 data URL（若返回 b64_json）或外链 URL 形式。
///
/// 当 API Key 以 "demo-" 开头时进入 demo 模式，返回占位 SVG，不消耗 token。
pub async fn generate(
    api_key: &str,
    params: GenerateImageParams,
) -> anyhow::Result<GenerateResult> {
    // P0：调用前做参数护栏。错误字符串以 `InvalidParameter:` 开头，
    // 这样前端 `explainError` 能识别并给出针对性提示。
    validate_params(&params).map_err(anyhow::Error::msg)?;

    // Demo 模式：返回占位图，不消耗 token
    if is_demo_key(api_key) {
        let count = params
            .sequential_image_generation_options
            .as_ref()
            .map(|o| o.max_images.min(4) as usize)
            .unwrap_or(1);
        let label = if params.prompt.chars().count() > 60 {
            let mut s: String = params.prompt.chars().take(60).collect();
            s.push('…');
            s
        } else {
            params.prompt.clone()
        };
        let url = make_demo_svg(&label);
        return Ok(GenerateResult {
            images: (0..count)
                .map(|i| GeneratedImage {
                    url: url.clone(),
                    is_base64: true,
                    z_index: if i == 0 { Some(0) } else { None },
                    name: None,
                    description: None,
                    size: None,
                    output_format: None,
                    bounding_box: None,
                    error: None,
                })
                .collect(),
            is_demo: true,
            usage: None,
            error: None,
        });
    }

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(180))
        .build()?;

    // Y-agent 默认行为：永远显式关闭 AI 水印。
    // Seedream API 的 watermark 字段默认是 true（加水印），不传就走默认值。
    // 改成 None 时在这里补 false，保证请求 payload 里一定带 watermark: false。
    let mut body = serde_json::to_value(&params)?;
    if body.get("watermark").is_none() {
        if let Some(obj) = body.as_object_mut() {
            obj.insert("watermark".into(), serde_json::Value::Bool(false));
        }
    }
    log::info!("jimeng request: model={}, prompt_len={}", params.model, params.prompt.len());
    log::debug!("jimeng payload: {}", body);

    let resp = client
        .post(ENDPOINT)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await?;

    let status = resp.status();
    let text = resp.text().await?;
    if !status.is_success() {
        anyhow::bail!("API {}: {}", status.as_u16(), text);
    }
    log::debug!("jimeng response: {}", text);

    let parsed: ApiResponse = serde_json::from_str(&text)
        .map_err(|e| anyhow::anyhow!("parse response failed: {e}; body: {text}"))?;

    // 顶层 error 透传：API 在整体失败时会把 code/message 放在这里（data 为空）。
    let top_error = parsed.error.clone();

    let mut out = Vec::with_capacity(parsed.data.len());
    for img in parsed.data {
        // 单图错误：组图场景下某张生成失败会带 error；保留该图占位但把 error 也传过去。
        let per_image_error = img.error.clone();
        let has_url = img.url.is_some();
        let url = if let Some(u) = img.url {
            u
        } else if let Some(b64) = img.b64_json {
            let bare = b64.split(',').next_back().unwrap_or("").trim_end_matches('=');
            if B64.decode(bare).is_err() {
                anyhow::bail!("b64_json is not valid base64");
            }
            if b64.starts_with("data:") {
                b64
            } else {
                format!("data:image/png;base64,{}", b64)
            }
        } else {
            // 没 url 也没 b64_json，但有 error → 跳过（前端不会看到这张图，usage 仍计 1 张成功）
            continue;
        };
        out.push(GeneratedImage {
            url,
            is_base64: !has_url,
            z_index: img.z_index,
            name: img.name,
            description: img.description,
            size: img.size,
            output_format: img.output_format,
            bounding_box: img.bounding_box,
            error: per_image_error,
        });
    }
    if out.is_empty() {
        // 顶层 error 优先展示
        if let Some(e) = top_error {
            let code = e.code.unwrap_or_default();
            let msg = e.message.unwrap_or_default();
            anyhow::bail!("API error ({code}): {msg}");
        }
        anyhow::bail!("API returned no images");
    }
    Ok(GenerateResult {
        images: out,
        is_demo: false,
        usage: parsed.usage,
        error: top_error,
    })
}
